docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,557 @@
+
+</br>
+
+In this approach, we will rely on `peerTracker`'s scoring mechanism that tracks the fastest connected nodes, and blocks malicious ones. To make use of this functionality, we track two important internal events from `peerTracker` in `libhead.Exchange` for peers selection and persistence purposes, and that is by proposing a design that allows access to `peerTracker`'s list of good peers and bad peers, and opening the design space to integrate peer selection and storage logic.
+
+## Implementation
+
+</br>
+
+### What systems will be affected?
+
+* `peerTracker` and `libhead.Exchange` from `go-header`
+* `newInitStore` in `nodebuilder/header.go`
+
+### Storage
+
+We will use a badgerDB datastore to store the addresses of previously seen good peers. The database will be stored in the `data` directory, and will have a store prefix `good_peers`. The database will have a single key-value pair, where the key is `peers` and the value is a `[]PeerInfo`
+
+The peer store will implement the following interface:
+
+```go
+type PeerStore interface {
+   // Put adds a list of peers to the store.
+    Put([]peer.AddrInfo) error
+    // Get returns a peer from the store.
+    Load() ([]peer.AddrInfo, error)
+    // Delete removes a peer from the store.
+    Delete(peer.ID) error
+}
+```
+
+\*: _Delete will be used in in the special case where we store a peer and then it gets blocked by `peerTracker`_
+
+And we expect the underlying implementation to use a badgerDB datastore. Example:
+
+```go
+type peerStore struct {
+    db datastore.Datastore
+}
+```
+
+### Peer Selection & Persistence
+
+To periodically select the _"good peers"_ that we're connected to and store them in the peer store, we will rely on the internals of `go-header`, specifically the `peerTracker` and `libhead.Exchange` structs.
+
+`peerTracker` has internal logic that continuiously tracks and garbage collects peers based on whether they connected to the node, disconnected from the node, or responded to the node while taking into consideration response speed. All connected peers are kept track of in a list, and disconnected peers are marked for removal, all as a part of a garbage collection routine. `peerTracker` also blocks peers that behave maliciously, but not in a routine, instead, it does so on `blockPeer` call.
+
+We intend to use this internal logic to periodically select peers that are "good" and store them in the peer store mentioned in the storage section. To do this, we will "hook" into the internals of `peerTracker` and `libhead.Exchange` by defining new "event handlers" intended to handle two main events from `peerTracker`:
+
+1. `OnUpdatedPeers`: _This event is triggered on every garbage collection cycle from `peerTracker`. Before the triggering of this event, the list would have underwent changes from other routines as well and thus the most final version will be available at the time of triggering this event ._
```

Comment:
This hook should not be called OnUpdatedPeers if it is triggered on every gc cycle. Probably should be called OnGCPeers or OnPeerGC

Reply:
Although `OnUpdatedPeers` seems to be a misnomer because it's an event that happens within GC, it's still descriptive as to what actually happens to the peer list. `OnGCPeers` sounds a bit vague to me tbh

Reply:
Never name things for how they are used elsewhere, name them for what they are